File: Sources/DanmakuOverlay/Time/TimelineFormatter.swift
import Foundation

enum TimelineFormatter {
    static func string(from seconds: Double) -> String {
        let safe = seconds.isFinite ? max(seconds, 0) : 0
        let totalSeconds = safe >= Double(Int.max) ? Int.max : Int(safe.rounded(.down))
        let hours = totalSeconds / 3_600
        let minutes = totalSeconds % 3_600 / 60
        let remainingSeconds = totalSeconds % 60

        if hours > 0 {
            return String(format: "%d:%02d:%02d", hours, minutes, remainingSeconds)
        }
        return String(format: "%02d:%02d", minutes, remainingSeconds)
    }
}

File: Windows/src/DanLaiDanQu.Core/TimelineFormatter.cs
namespace DanLaiDanQu.Core;

public static class TimelineFormatter
{
    public static string Format(double seconds)
    {
        var safe = double.IsFinite(seconds) ? Math.Max(seconds, 0) : 0;
        var totalSeconds = safe >= long.MaxValue ? long.MaxValue : (long)Math.Floor(safe);
        var hours = totalSeconds / 3_600;
        var minutes = totalSeconds % 3_600 / 60;
        var remainingSeconds = totalSeconds % 60;

        return hours > 0
            ? $"{hours}:{minutes:00}:{remainingSeconds:00}"
            : $"{minutes:00}:{remainingSeconds:00}";
    }
}
